File: cpp/src/0083.cpp
/*
[83. Remove Duplicates from Sorted List (Easy)](https://leetcode.com/problems/remove-duplicates-from-sorted-list/)

@topic: Linked List

Constraints:
  - The number of nodes in the list is in the range `[0, 300]`.
  - `-100 <= Node.val <= 100`
  - The list is guaranteed to be **sorted** in ascending order.
*/

#include "structs/LinkedList.hpp"

class Solution {
   public:
    ListNode* deleteDuplicates(ListNode* head) {
        if (!head || !head->next) return head;
        auto curr = head;
        while (curr->next) {
            auto next = curr->next;
            if (curr->val == next->val) {
                curr->next = next->next;
                delete next;
            } else {
                curr = curr->next;
            }
        }
        return head;
    }
};

#include <cassert>
#include <chrono>
#include <iostream>
#include <vector>

using std::tuple;
using std::vector;
using std::chrono::duration_cast;
using std::chrono::microseconds;
using std::chrono::system_clock;

int main() {
    using namespace LinkedList;

    Solution o;

    vector<tuple<ListNode*, ListNode*>> CASES = {
        {of({1, 1, 2}), of({1, 2})},
        {of({1, 1, 2, 3, 3}), of({1, 2, 3})},
    };

    for (auto& [head, excepted] : CASES) {
        assert(eq(o.deleteDuplicates(head), excepted));
    }

    auto start = system_clock::now();
    for (auto i = 0; i < 100000; ++i) {
        o.deleteDuplicates(of({1, 1, 2}));
        o.deleteDuplicates(of({1, 1, 2, 3, 3}));
    }
    auto end = system_clock::now();
    std::cout << duration_cast<microseconds>(end - start).count() << std::endl;
}

File: python/src/0083.py
from typing import Optional

from structs.ListNode import ListNode, create


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def deleteDuplicates(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head or not head.next:
            return head
        walk = head
        while walk.next:
            if walk.val == walk.next.val:
                walk.next = walk.next.next
            else:
                walk = walk.next
        return head


if __name__ == '__main__':
    import unittest

    CASES = (
        (create(1, 1, 2), create(1, 2)),
        (create(1, 1, 2, 3, 3), create(1, 2, 3)),
    )

    class Test(unittest.TestCase):
        def test(self):
            deleteDuplicates = Solution().deleteDuplicates
            for head, excepted in CASES:
                actual = deleteDuplicates(head)
                self.assertEqual(actual, excepted)

    unittest.main()
